private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,269 @@
+	// is fine; they are almost never written to so false sharing is not a
+	// concern.
+	initParams atomic.Pointer[protocol.InitializeParams]
+	traceValue atomic.Pointer[protocol.TraceValue]
+}
+
+// init performs *actual* initialization of the server. This is called by Initialize().
+//
+// It may only be called once for a given server.
+func (l *lsp) init(params *protocol.InitializeParams) error {
+	if l.initParams.Load() != nil {
+		return fmt.Errorf("called the %q method more than once", protocol.MethodInitialize)
+	}
+	l.initParams.Store(params)
+
+	// TODO: set up logging. We need to forward everything from server.logger through to
+	// the client, if tracing is turned on. The right way to do this is with an extra
+	// goroutine and some channels.
+
+	return nil
+}
+
+// findImportable finds all files that can potentially be imported by the proto file at
+// uri. This returns a map from potential Protobuf import path to the URI of the file it would import.
+//
+// Note that this performs no validation on these files, because those files might be open in the
+// editor and might contain invalid syntax at the moment. We only want to get their paths and nothing
+// more.
+func (l *lsp) findImportable(
+	ctx context.Context,
+	uri protocol.URI,
+) (map[string]bufimage.ImageFileInfo, error) {
+	fileInfos, err := l.controller.GetImportableImageFileInfos(ctx, uri.Filename())
+	if err != nil {
+		return nil, err
+	}
+
+	imports := make(map[string]bufimage.ImageFileInfo)
+	for _, fileInfo := range fileInfos {
+		imports[fileInfo.Path()] = fileInfo
+	}
+
+	l.logger.Sugar().Debugf("found imports for %q: %#v", uri, imports)
+
+	return imports, nil
+}
+
+// newHandler constructs an RPC handler that wraps the default one from jsonrpc2. This allows us
+// to inject debug logging, tracing, and timeouts to requests.
+func (l *lsp) newHandler() jsonrpc2.Handler {
+	actual := protocol.ServerHandler(newServer(l), nil)
+	return func(ctx context.Context, reply jsonrpc2.Replier, req jsonrpc2.Request) (retErr error) {
+		ctx, span := l.tracer.Start(
+			ctx,
+			tracing.WithErr(&retErr),
+			tracing.WithAttributes(attribute.String("method", req.Method())),
+		)
+		defer span.End()
+
+		l.logger.Debug(
+			"processing request",
+			zap.String("method", req.Method()),
+			zap.ByteString("params", req.Params()),
+		)
+
+		ctx = withRequestID(ctx)
+
+		replier := l.adaptReplier(reply, req)
+
+		// Verify that the server has been initialized if this isn't the initialization
+		// request.
+		if req.Method() != protocol.MethodInitialize && l.initParams.Load() == nil {
+			return replier(ctx, nil, fmt.Errorf("the first call to the server must be the %q method", protocol.MethodInitialize))
+		}
+
+		return actual(ctx, replier, req)
+	}
+}
+
+// adaptReplier wraps a jsonrpc2.Replier, allowing us to inject logging and tracing and so on.
+func (l *lsp) adaptReplier(reply jsonrpc2.Replier, req jsonrpc2.Request) jsonrpc2.Replier {
```

Comment:
I think the preferred naming convention in the code base for wrapping/wrapped types tend to be `wrapReplier` and `wrappedConnection` for the type below.